assets/hip-206/solidity/IHederaTokenService.sol

```diff
@@ -160,8 +160,12 @@ interface IHederaTokenService {
         uint32 amount;
 
         // The denomination of the fee; taken as hbar if left unset and taken as the id
-        // of the newly created token if set to the sentinel value of 0x0
+        // of the newly created token if set to the sentinel value of 0x0 (since there
+        // is no notion of unset or null value in Solidity and the default value of an
+        // address variable is 0x0, the `isTokenIdSet` variable is used to distinguish
+        // between the two cases)
         address tokenId;
+        bool isTokenIdSet;
```

Comment:
Why would tokenId ever be un-set?  If passed in as an argument I think it's reasonable to expect that the required fields are set.  Is there any meaning to a fixed fee with an amount, a fee collector, and an unset tokenI?

Reply:
Yes, there are some special meanings in the protobufs values 0 and unset for this field:

- if tokenId is 0, then use the token that's being created in this tx's id
- if tokenId is unset, then use hbars
- if tokenId is set to non-zero number, then use the id specified

When decoding Solidity input args however it's impossible to distinguish between cases 1 and 2 since missing value and zero value are both decoded to 0x0.

I do think that we can make it easier to read and understand in the Solidity struct though, maybe something like:

address tokenId;
bool useHbarsForPayment;
bool useCurrentTokenForPayment;

and specify that exactly one of the fields must be populated. If more than one field is populated we would fail the decoding, set error message in the message frame and return null from the precompile invocation.